In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import sqlite3
import seaborn as sns

To Import Data we will first define a function which will simplify importation

In [340]:
def import_data (file_path, delimiter=',',encoding=None):
    if delimiter == ',':
        return pd.read_csv(file_path,encoding=encoding)
    elif delimiter == '\t':
        return pd.read_csv(file_path,delimiter='\t',encoding=encoding)
    else:
        return print("You are using an invalid delimiter")


In [ ]:
imdb_path = r'C:\Users\espym.LAPTOP-41F90NSA\Desktop\zippedData\im.db\im.db'
conn = sqlite3.connect(imdb_path)


In [341]:
file_path1 = r'C:\Users\espym.LAPTOP-41F90NSA\projects\Movie_Project-1\tmdb.movies.csv.gz'
tmdbDF = import_data(file_path1)

In [342]:
file_path2 = r'C:\Users\espym.LAPTOP-41F90NSA\projects\Movie_Project-1\tn.movie_budgets.csv.gz'
tn_movieDF = import_data(file_path2)

In [343]:
file_path3 = r'C:\Users\espym.LAPTOP-41F90NSA\projects\Movie_Project-1\rt.movie_info.tsv.gz'
movie_infoDF = import_data(file_path3, delimiter='\t')

In [344]:
file_path4 =r'C:\Users\espym.LAPTOP-41F90NSA\projects\Movie_Project-1\rt.reviews.tsv.gz'
reviewsDF = import_data(file_path4, delimiter='\t', encoding='latin-1')

## Preview of Information in the data

Here we just want to have a look at whats contained in the IMDB dataset and just have a proper feel of whats contained in the dataset

In [346]:
#from the imdb database we first want to know the names and number of the tables.
query0 = ("""
    SELECT name
    AS 'Table_Name'
    FROM sqlite_master
    WHERE type = 'table'
;
""")
pd.read_sql(query0,conn)

In [363]:
# checking for the information contained in the movie basics table
query1 = ("""
    SELECT * 
    FROM movie_basics
    ;
          """)
df = pd.read_sql(query1,conn)
df.head()

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

To get the preview of information from the data in a simplified way, we will define another function that will help do so

In [368]:
def dataframe_preview(df):
    # #To get the first five rows
    # print("The first five rows:")
    # print(df.head())
    # print("\n")
    # #To get the last five rows
    # print("The last five rows:")
    # print(df.tail())
    # print('\n')
    #To get the shape of the dataframe
    print("The shape of the dataframe:")
    print(df.shape)
    print('\n')
    #To get the info of the dataframe
    print("The data in the dataframe:")
    print(df.info())
    print('\n')
    #To get the descriptive statistics of the dataframe
    print("The descriptive stats of the df:")
    print(df.describe())
    #To get the columns of the dataframe
    print("The columns of the dataframe are:")
    print(df.columns)


In [369]:
dataframe_preview(df)

The shape of the dataframe:
(146144, 6)


The data in the dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   movie_id         146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB
None


The descriptive stats of the df:
          start_year  runtime_minutes
count  146144.000000    114405.000000
mean     2014.621798        86.187247
std         2.733583       166.360590
min      2010.000000         1.000000
25%      2012.000000        70.000000
50%      2015.000000        87.000000
75%      2017.000000        99.000000
max      2115.000000     51420.000000
T

Now that we have a feel of the movie_basics, well also explore movie_ratings table

In [370]:
query2 = ("""
    SELECT * 
    FROM movie_ratings
    ;
        """)
df2 = pd.read_sql(query2,conn)
df2.head()

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

In [371]:
#We will call the function dataframe preview to view the info in the dataframe
dataframe_preview(df2)

The shape of the dataframe:
(73856, 3)


The data in the dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   movie_id       73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB
None


The descriptive stats of the df:
       averagerating      numvotes
count   73856.000000  7.385600e+04
mean        6.332729  3.523662e+03
std         1.474978  3.029402e+04
min         1.000000  5.000000e+00
25%         5.500000  1.400000e+01
50%         6.500000  4.900000e+01
75%         7.400000  2.820000e+02
max        10.000000  1.841066e+06
The columns of the dataframe are:
Index(['movie_id', 'averagerating', 'numvotes'], dtype='object')


In [364]:
#Now to preview info from the last dataset we will be using
dataframe_preview(tn_movieDF)

The shape of the dataframe:
(5782, 6)


The data in the dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB
None


The descriptive stats of the df:
                id
count  5782.000000
mean     50.372363
std      28.821076
min       1.000000
25%      25.000000
50%      50.000000
75%      75.000000
max     100.000000


In [365]:
# Checking the first five rows of the dataset
tn_movieDF.head()

id  release_date                                        movie  \
0   1  Dec 18, 2009                                       Avatar   
1   2  May 20, 2011  Pirates of the Caribbean: On Stranger Tides   
2   3   Jun 7, 2019                                 Dark Phoenix   
3   4   May 1, 2015                      Avengers: Age of Ultron   
4   5  Dec 15, 2017            Star Wars Ep. VIII: The Last Jedi   

  production_budget domestic_gross worldwide_gross  
0      $425,000,000   $760,507,625  $2,776,345,279  
1      $410,600,000   $241,063,875  $1,045,663,875  
2      $350,000,000    $42,762,350    $149,762,350  
3      $330,600,000   $459,005,868  $1,403,013,963  
4      $317,000,000   $620,181,382  $1,316,721,747

## Data Cleaning

Now that we have a feeling of the data, we will now combine the two tables i.e movie_basics and movie_ratings using their primary key movie_id

In [372]:
#Joining the tables using the movie_id column
query3 = ("""
SELECT movie_id, primary_title,start_year,runtime_minutes,genres,averagerating AS 'movie_rating'
FROM movie_basics
JOIN movie_ratings
USING (movie_id)
;
""")
df = pd.read_sql(query3,conn)

In [373]:
dataframe_preview(df)

The shape of the dataframe:
(73856, 6)


The data in the dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   movie_id         73856 non-null  object 
 1   primary_title    73856 non-null  object 
 2   start_year       73856 non-null  int64  
 3   runtime_minutes  66236 non-null  float64
 4   genres           73052 non-null  object 
 5   movie_rating     73856 non-null  float64
dtypes: float64(2), int64(1), object(3)
memory usage: 3.4+ MB
None


The descriptive stats of the df:
         start_year  runtime_minutes  movie_rating
count  73856.000000     66236.000000  73856.000000
mean    2014.276132        94.654040      6.332729
std        2.614807       208.574111      1.474978
min     2010.000000         3.000000      1.000000
25%     2012.000000        81.000000      5.500000
50%     2014.000000        91.000000      6.5000

We will now go ahead and clean the dataset to remove the missing values in the dataframe. The columns runtime_minutes and genres appear to having missing values

In [378]:
#checking for the sum of null values
null_count = df.isnull().sum()
null_count

movie_id              0
primary_title         0
start_year            0
runtime_minutes    7620
genres              804
movie_rating          0
dtype: int64

In [379]:
#checking for all the values
non_null_count = df.shape[0]
non_null_count

73856

In [380]:
#checking the percentage of the null values to determine if to drop them
null_percentage = (null_count / non_null_count) * 100
print(null_percentage)

movie_id            0.000000
primary_title       0.000000
start_year          0.000000
runtime_minutes    10.317374
genres              1.088605
movie_rating        0.000000
dtype: float64


In [381]:
#The null values take a minimal percentage of the total values and can be droppped
df.dropna(inplace=True)
df.shape

(65720, 6)

In [383]:
#checking null values in the other dataframe
budgetDF = tn_movieDF
budgetDF.isnull().sum()

id                   0
release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64

The budget dataframe contains no null values

Conversion of datatypes

The prodution_budget,domestic_gross and worldwide_gross are denoted as objects and we need them as float in order to mathematical computations


In [384]:
budgetDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


In [ ]:
#Conversion of The prodution_budget,domestic_gross and worldwide_gross to float
budgetDF['production_budget'] = budgetDF['production_budget'].str.replace('$', '').str.replace(',', '').astype(float)
budgetDF['domestic_gross'] = budgetDF['domestic_gross'].str.replace('$', '').str.replace(',', '').astype(float)
budgetDF['worldwide_gross'] = budgetDF['worldwide_gross'].str.replace('$', '').str.replace(',', '').astype(float)

Checking for duplicates 


In [ ]:
df.duplicated().sum()

In [ ]:
budgetDF.duplicated().sum()

From both the datasets, there are no duplicated values

Checking for data consistency

In [ ]:
df.describe()

From this we have a few issues in regards to runtime
1. The minimum runtime is 3 minutes which is quite low to meet a movie threshold
2. The maximum runtime is 51420 which an extreme value

filtering the values to include a minimum threshold of 20 minutes and 180 minutes 

In [ ]:
clean_runtime = df[(df['runtime_minutes'] >= 20) & (df['runtime_minutes'] <= 180)]
sum_of_movies_within_threshold = len(clean_runtime)
print('The movies within the threshold are:', sum_of_movies_within_threshold)

In [ ]:
df = df[(df['runtime_minutes'] >= 20) & (df['runtime_minutes'] <= 200)]

In [ ]:
df.describe()

In [ ]:
budgetDF.info()

We have an issue with the release date as its an object and its needed to be in date format

In [ ]:
#conversion of release date from object to datetime
budgetDF['release_date'] = pd.to_datetime(budgetDF['release_date'])

In [ ]:
budgetDF.describe()

From this we also have the issue of domestic gross and worldwide gross are zero 

In [ ]:
#checking for the duration of info in the budget dataframe
start_date = budgetDF['release_date'].min()
end_date = budgetDF['release_date'].max()

print(f'the start date is {start_date} and the end date is {end_date}')


From this we see the first date is 1915 which is significantly a long time ago and the film and production industry has changed in one way or another. We will now convert the dataframe to start in the the year 2000 and end in 2019 as 2020 has no gross earnings

In [ ]:
start_date = pd.to_datetime('2000-1-1 00:00:00')
end_date = pd.to_datetime('2019-12-31 00:00:00')
budgetDF = budgetDF[(budgetDF['release_date'] >= start_date) & (budgetDF['release_date'] <= end_date)]

## Explarotary Analysis

In [ ]:
#checking for the year and duration of the release dates 
start_date = df['start_year'].min()
end_date = df['start_year'].max()
period_taken = end_date - start_date

print('start:', start_date)
print('stop:', end_date)
print('durarion:', period_taken,'years')

In [ ]:
#checking for the duration of info in the budget dataframe
start_date = budgetDF['release_date'].min()
end_date = budgetDF['release_date'].max()
period_taken = end_date -  start_date

print('The start date is', start_date)
print('The end date id', end_date)
print('Period of the dataset is', period_taken)

Univariate Analysis

In [ ]:
df.columns

In [ ]:
df['primary_title'].describe()

In [ ]:
#Checking for the most common genre
df['genres'].describe()


In [ ]:
df['genres'].value_counts()

In [ ]:
budgetDF.columns

In [ ]:
budgetDF['movie'].describe()

2.1 Bivariate Analysis

2.1.1 How has the average movie runtime changed over the years

In [ ]:
df.columns

In [ ]:
runtime_changes_over_years = df.groupby('start_year')['runtime_minutes'].mean()
print(runtime_changes_over_years)

In [ ]:
#To visualize the average runtime through the years
plt.figure(figsize=(12,10))
sns.barplot(x='start_year', y='runtime_minutes', data=df)
plt.title('average running time in years')
plt.show()

2.1.2 Analyzing the distribution between Genres and Runtime

In [ ]:
df.columns

In [ ]:
#Checking for the the first 10 movies with the highest runtimes and their genres
top_10_runtime = df.nlargest(10,'runtime_minutes')[['primary_title','runtime_minutes','genres']].set_index('primary_title')
top_10_runtime

In [ ]:
#To visualize the distrubution of runtimes and their genres
plt.figure(figsize=(16,10))
df['genre'] = df['genres'].str.split(',').str[0].str.strip()
sns.boxplot(data=df, x='genre', y='runtime_minutes', palette='Set2')
plt.xlabel('Genre')
plt.ylabel('Runtime (minutes)')
plt.title('Box Plot of Runtimes by Genre')
plt.xticks(rotation=45)
plt.show()


2.1.3 What is the average movie rating over the years

In [ ]:
df.columns

In [ ]:
average_ratings_over_years = df.groupby('start_year')['movie_rating'].mean().sort_values()
print(average_ratings_over_years)

In [ ]:
plt.figure(figsize=(16,10))
sns.barplot(x='start_year',y='movie_rating',data=df)
plt.title('Average rating through the years')
plt.show()

2.1.4 Checking for the distribution of ratings and genres

In [ ]:
#Checking for the the first 10 movies with the highest runtimes and their genres
top_10_ratings = df.nlargest(10,'movie_rating')[['primary_title','movie_rating','genres']].set_index('primary_title')
top_10_ratings

From the top 10 top rated movies, documentaries is the most common genre

In [ ]:
#To visualize the relationship between genres and their ratings
plt.figure(figsize=(16,10))
df['genre'] = df['genres'].str.split(',').str[0].str.strip()
sns.boxplot(data=df, x='genre', y='movie_rating', palette='Set2')
plt.xlabel('Genre')
plt.ylabel('Movie Ratings')
plt.title('Box Plot of Runtimes by Genre')
plt.xticks(rotation=45)
plt.show()

2.1.5 Checking for the year with the highest average gross earnings

In [ ]:
budgetDF.columns

The release date is inclusive of date and month and for this we just want the years. We will convert the release date to years only

In [ ]:
#conversion to years
budgetDF['release_date'] = budgetDF['release_date'].dt.year

In [ ]:
average_earnings = budgetDF.groupby('release_date')['worldwide_gross'].mean().sort_values(ascending=False)
print(average_earnings)

In [ ]:
#To visualize the average earnings throughout the year
plt.figure(figsize=(12,6))
sns.barplot(x='release_date', y='worldwide_gross',data=budgetDF)
plt.title('Average Worldwide Earnings 2000 - 2020')
plt.show()

2.1.6 Checking for correlation tp between budget and worldwide income

In [ ]:
budgetDF.columns

In [ ]:
plt.figure(figsize=(16,10))
sns.scatterplot(x='production_budget', y='worldwide_gross', data=budgetDF)
plt.title('relationship btwn budget and revenue')
plt.tight_layout()
plt.show()
